### PR TITLE
feat(noise): add Noise Handshake State Machine and primitives

### PR DESCRIPTION
**Pull request by s1fr0**
This PR introduces data structures and procedures required to process Noise Handshake message patterns according to [Noise specification](http://www.noiseprotocol.org/noise.html#processing-rules). 

Specifically, it adds:
- Definition of the Noise Handshake `K1K1`, `XK1`, `XX`, `XXpsk0`
- Definition of Waku Payload protocol IDs as per [35/WAKU2-NOISE](https://rfc.vac.dev/spec/35/#protocol-payload-format) 
- [Noise state machine](http://www.noiseprotocol.org/noise.html#overview-of-handshake-state-machine) (Noise Handshake State, Symmetric State and Cipher State). 
- Noise processing rules primitives for states (names from Noise specification):
    - Handshake State: `InitializeKey`, `HasKey`, `EncryptWithAd`, `DecryptWithAd`;
    - Symmetric State: `InitializeSymmetric`, `MixKey`, `MixHash`, `MixKeyAndHash`, `EncryptAndHash`, `DecryptAndHash`, `Split`;
    - Handshake State: `Initialize`. 
- Other utility functions (`genKeyPair` for generating Curve25519 key pairs, `print` for printing handshake patterns, `hashProtocol` for initializing handshake states, `dh` to perform Diffie-Hellman operations)

The implementation of the Noise state machine and related primitives follows the Noise state machine implementation of [noise-libp2p](https://github.com/status-im/nim-libp2p/blob/master/libp2p/protocols/secure/noise.nim).

This PR extends https://github.com/status-im/nim-waku/pull/932 (which in turn extends https://github.com/status-im/nim-waku/pull/931 and https://github.com/status-im/nim-waku/pull/930) and partially addresses https://github.com/status-im/nim-waku/issues/881.

### CODE REVIEW

**Comment by staheri14** [comment]
Can you please include the summary of changes in the PR? it is hard to assess the correctness of the implementation without knowing the expected outcome of this PR.

**Comment by s1fr0** [comment]
> Can you please include the summary of changes in the PR? it is hard to assess the correctness of the implementation without knowing the expected outcome of this PR. The same for other Noise PRs, thanks!

Sure, done! Thanks!

**Comment by staheri14** [comment]
I think this one also builds on top of #932, so I'll wait for that to get merged before leaving a review.

**Comment by s1fr0** [comment]
@staheri14 @jm-clius ready for review. Thanks!

File: waku/v2/protocol/waku_noise/noise.nim [hunk 1]
```diff
@@ -62,6 +77,94 @@ type
     nonce: ChaChaPolyNonce
     ad: seq[byte]
 
+  #################################
+  # Noise handshake patterns
+  #################################
+
+  # The Noise tokens appearing in Noise (pre)message patterns
```
Comment (staheri14) [inline waku/v2/protocol/waku_noise/noise.nim:84]: Do we use the same term in our own spec i.e., pre-message? if not, I'd suggest using consistent terms

File: waku/v2/protocol/waku_noise/noise.nim [hunk 2]
```diff
@@ -98,10 +261,276 @@ proc genKeyPair*(rng: var BrHmacDrbgContext): KeyPair =
   keyPair.publicKey = keyPair.privateKey.public()
   return keyPair
 
+# Prints Handshake Patterns using Noise pattern layout
+proc print*(self: HandshakePattern) 
+   {.raises: [IOError, NoiseMalformedHandshake].}=
+  try:
+    if self.name != "":
+      echo self.name, ":"
```
Comment (staheri14) [inline waku/v2/protocol/waku_noise/noise.nim:269]: An `echo` is left behind!

**Review by staheri14 — COMMENTED** [review]
Thanks for the PR and great documentation! would you please explain how would you use these primitives in the final implementation (at a very high level, just saying that how each func gets called in a sample handshake scenario)?  while the description of each proc indicates what it does, it is hard to grasp how they fit in the big picture. Do you have an integration test for these primitives?

File: waku/v2/protocol/waku_noise/noise.nim (hunk 2 above)
Comment (s1fr0) [reply waku/v2/protocol/waku_noise/noise.nim:269]: It was a quick way to have a new line at the end! I made it consistent with a `stdout.write`

**Comment by s1fr0** [comment]
@staheri14 Added a short overview in https://github.com/status-im/nim-waku/commit/9047e8cd9d03261e1bc6943bb8db702169ccc61e, I think that adding more details will essentially duplicate Noise specifications. 

Regarding integration tests for these primitives, these are present in https://github.com/status-im/nim-waku/pull/934, where state machine primitives will be consistently checked during a full handshake execution and in after-handshake. The reason why tests are there is because, originally, this and https://github.com/status-im/nim-waku/pull/934 were a single PR. When were splitted, I decided to keep implementation for handshake processing rules and all handshake-related tests in same PR.

However, as per description, Noise state machine primitives integrated by this PR mainly correspond to the ones provided in  [libp2p-noise](https://github.com/status-im/nim-libp2p/blob/master/libp2p/protocols/secure/noise.nim) which are already tested [here](https://github.com/status-im/nim-libp2p/blob/master/tests/testnoise.nim).

File: waku/v2/protocol/waku_noise/noise.nim [hunk 3]
```diff
@@ -98,10 +261,288 @@ proc genKeyPair*(rng: var BrHmacDrbgContext): KeyPair =
...
+# Prints Handshake Patterns using Noise pattern layout
+proc print*(self: HandshakePattern) 
+   {.raises: [IOError, NoiseMalformedHandshake].}=
+  try:
+    if self.name != "":
+      stdout.write self.name, ":\n"
+      stdout.flushFile()
+    #We iterate over pre message patterns, if any
+    if self.preMessagePatterns != EmptyPreMessagePattern:
+      for pattern in self.pre_messagePatterns:
```
Comment (jm-clius) [inline waku/v2/protocol/waku_noise/noise.nim:273]:
```suggestion
      for pattern in self.preMessagePatterns:
```

File: waku/v2/protocol/waku_noise/noise.nim [hunk 4]
```diff
@@ -98,10 +261,288 @@ proc genKeyPair*(rng: var BrHmacDrbgContext): KeyPair =
   keyPair.publicKey = keyPair.privateKey.public()
   return keyPair
 
+# Prints Handshake Patterns using Noise pattern layout
+proc print*(self: HandshakePattern)
```
Comment (jm-clius) [inline waku/v2/protocol/waku_noise/noise.nim:265]: Out of curiosity, what is the benefit of having a `stdout` print utility rather than encoding this in a `'$'` stringify procedure, which can be used for more than just stdout? Guessing it may be too complex?

**Review by jm-clius — COMMENTED** [review]
Thanks! LGTM, except that the diff contains all these submodules, which I think was unintentional? You may need to run `make clean update` in your local dev environment to ensure your submodules are up to date. Also a good idea to check the diff before you open a PR to ensure there are no changes included unrelated to that PR.

**Comment by s1fr0** [comment]
@jm-clius Thanks a lot for clarifying this to me. I actually wrongly assumed the opposite, that is that I had updated submodules locally and that's why I pushed the diff for them. But thinking a bit more on that, I would have quickly realized to be wrong. Thanks!

File: waku/v2/protocol/waku_noise/noise.nim [hunk 5]
```diff
@@ -62,6 +77,94 @@ type
     nonce: ChaChaPolyNonce
     ad: seq[byte]
 
+  #################################
+  # Noise handshake patterns
+  #################################
+
+  # The Noise tokens appearing in Noise (pre)message patterns 
+  # as in http://www.noiseprotocol.org/noise.html#handshake-pattern-basics
+  NoiseTokens* = enum
```
Comment (jm-clius) [inline waku/v2/protocol/waku_noise/noise.nim:86]: Assuming these types are exported `*` to be unit tested or used in other modules?

Comment (s1fr0) [reply waku/v2/protocol/waku_noise/noise.nim]: Good catch! There is no need to export the tokens as long as new handshake patterns are defined in this module. Removed in https://github.com/status-im/nwaku/pull/933/commits/ed3eff8855f531451f5003c14bea9db237dfa4bb

File: tests/v2/test_waku_noise.nim [hunk 6]
```diff
@@ -157,4 +158,180 @@ procSuite "Waku Noise":
...
+      ck != getChainingKey(symmetricState)
+      h != getHandshakeHash(symmetricState)
+
+    # We update test values
+    cs = getCipherState(symmetricState)
+    ck = getChainingKey(symmetricState)
+    h = getHandshakeHash(symmetricState)
+
+    ########################################
+    # encryptAndHash and decryptAndHash
```
Comment (staheri14) [inline tests/v2/test_waku_noise.nim:283]: Other things to consider for the tests is the edge cases and unhappy paths, like when the nonce reaches the max value.

File: waku/v2/protocol/waku_noise/noise.nim [hunk 7]
```diff
@@ -98,10 +261,333 @@ proc genKeyPair*(rng: var BrHmacDrbgContext): KeyPair =
...
+  var authorizationTag: ChaChaPolyTag
+  ChaChaPoly.encrypt(state.k, nonce, authorizationTag, ciphertext, ad)
+
+  # We append the authorization tag to ciphertext
+  ciphertext.add(authorizationTag)
+
+  # We increase the Cipher state nonce
+  inc state.n
+  # If the nonce is greater than the maximum allowed nonce, we raise an exception
+  if state.n > NonceMax:
```
Comment (staheri14) [inline waku/v2/protocol/waku_noise/noise.nim:392]: Wondering why the `NonceMax+1` i.e., `uint64.high-1+1` value is an invalid nonce? it can still be used for the next message? I think the nonce needs to be check before increment i.e., to see if it is equal to `uint64.high`, and then avoid increment and throw an exception indicating the max value is hit. 
Ditto for the similar nonce max check in the `decryptWithAd` proc. https://github.com/status-im/nwaku/blob/795eefcd697cbae4e0039496c39544383682500a/waku/v2/protocol/waku_noise/noise.nim#L429

**Review by staheri14 — COMMENTED** [review]
Great code documentation and thanks for the unittests! The implementation looks good to me. I commented on some parts among which my major concern is about the nonce max value.

File: tests/v2/test_waku_noise.nim [hunk 8]
```diff
@@ -157,4 +158,180 @@ procSuite "Waku Noise":
...
+
+    # We generate random input key material and we execute mixKey
+    var inputKeyMaterial = randomChaChaPolyKey(rng[])
+    mixKey(symmetricState, inputKeyMaterial)
+
+    # mixKey changes the Symmetric State's chaining key and encryption key of the embedded Cipher State 
+    # It further sets to 0 the nonce of the embedded Cipher State
+    check:
+      getKey(cs) != getKey(getCipherState(symmetricState))
+      getNonce(getCipherState(symmetricState)) == 0.uint64
```
[the remark this answers is not in the document]
Comment (s1fr0) [reply tests/v2/test_waku_noise.nim:253]: Yes! Setting the nonce to zero is implicit in MixKey's call 

```
ss.cs = CipherState(k: tempKeys[1])
```

where a new Cipher State is created and only the key is assigned (hence, `n` goes to its default value 0). Probably better to make the assignment explicit.

File: waku/v2/protocol/waku_noise/noise.nim (hunk 7 above)
Comment (s1fr0) [reply waku/v2/protocol/waku_noise/noise.nim:392]: I agree with your observation! The reason is that, by specification, 2^64-1 is a reserved nonce. I quote verbatim:

> A CipherState responds to the following functions. The ++ post-increment operator applied to n means "use the current n value, then increment it". The maximum n value (2^64-1) is reserved for other use. If incrementing n results in 2^64-1, then any further EncryptWithAd() or DecryptWithAd() calls will signal an error to the caller.

more on [The Cipher State Object](http://www.noiseprotocol.org/noise.html#the-cipherstate-object).

Note that `NonceMax = uint64.high - 1`, that is `NonceMax = 2^64-1-1`. Hence, if `n > NonceMax` after an increment, it means that `n = NonceMax+1 = 2^64-1`. In fact, `NonceMax` is the last "valid" nonce that can be used in encryption/decryption according to specification.

File: waku/v2/protocol/waku_noise/noise.nim [hunk 9]
```diff
@@ -98,10 +261,333 @@ proc genKeyPair*(rng: var BrHmacDrbgContext): KeyPair =
...
+proc decryptAndHash*(ss: var SymmetricState, ciphertext: openArray[byte]): seq[byte]
+  {.raises: [Defect, NoiseDecryptTagError, NoiseNonceMaxError].} =
+  # The output plaintext
+  var plaintext: seq[byte]
+  # If an encryption key is set in the Symmetric state, we proceed with decryption using the handshake hash value as associated data
+  # Note that the ciphertext must contains an authorization tag, so its length should be greater or equal (in case of empty plaintext) the latter
+  if ss.cs.hasKey and ciphertext.len >= ChaChaPolyTag.len:
+    plaintext = ss.cs.decryptWithAd(ss.h.data, ciphertext)
+  # According to specification, if no key is set return the plaintext
+  else:
```
[the remark this answers is not in the document]
Comment (s1fr0) [reply waku/v2/protocol/waku_noise/noise.nim:541]: Returning the input in case no key is set is a Noise specific condition and not of ChaChaPoly encryption/decryption. 
I just realized that indeed the check `ciphertext.len >= ChaChaPolyTag.len` is superflous here: `decryptWithAd` will read the last `ChaChaPolyTag.len` bytes as

`  let inputAuthorizationTag = ciphertext.toOpenArray(ciphertext.len - ChaChaPolyTag.len, ciphertext.high).intoChaChaPolyTag`

In turn, toOpenArray while computing the slice, will read starting from `max(ciphertext.len - ChaChaPolyTag.len, 0)` which is 0 in case `ciphertext.len >= ChaChaPolyTag.len`. Hence the full ciphertext is processed by `intoChaChaPolyTag`, which has such [check](https://github.com/status-im/nim-libp2p/blob/9973b9466dd062d7406d0178c750249b7e48471f/libp2p/crypto/chacha20poly1305.nim#L49) already implemented.

File: waku/v2/protocol/waku_noise/noise.nim [hunk 10]
```diff
@@ -98,10 +261,333 @@ proc genKeyPair*(rng: var BrHmacDrbgContext): KeyPair =
...
+# Symmetric State primitives
+#################################
+
+# Initializes a Symmetric State
+proc init*(_: type[SymmetricState], hsPattern: HandshakePattern): SymmetricState =
+  var ss: SymmetricState
+  # We compute the hash of the protocol name
+  ss.h = hsPattern.name.hashProtocol
+  # We initialize the chaining key ck
+  ss.ck = result.h.data.intoChaChaPolyKey
```
[the remark this answers is not in the document]
Comment (s1fr0) [reply waku/v2/protocol/waku_noise/noise.nim:470]: Great catch! This bug was setting the initial chaining key to an empty value, rather than the handshake hash value (thus losing the initial binding with the protocol name). I fixed and added an extra check in unit tests to see if `h` and `ck` agree on their when initializing a Symmetric State.

File: waku/v2/protocol/waku_noise/noise.nim [hunk 11]
```diff
@@ -98,10 +261,333 @@ proc genKeyPair*(rng: var BrHmacDrbgContext): KeyPair =
...
+  ciphertext.add(plaintext)
+
+  # The nonce is read from the input CipherState
+  # By Noise specification the nonce is 4 bytes long out of the 12 bytes supported by ChaChaPoly
+  var nonce: ChaChaPolyNonce
+  nonce[4..<12] = toBytesLE(state.n)
+
+  # We perform encryption and we store the authorization tag
+  var authorizationTag: ChaChaPolyTag
+  ChaChaPoly.encrypt(state.k, nonce, authorizationTag, ciphertext, ad)
```
[the remark this answers is not in the document]
Comment (s1fr0) [reply waku/v2/protocol/waku_noise/noise.nim:384]: This is correct and it should be checked here as well rather than only in encryptAndHash/decryptAndHash. Since the latter always call encryptWithAd/decryptWithAd, respective, I moved the check there.

File: waku/v2/protocol/waku_noise/noise.nim [hunk 12]
```diff
@@ -62,6 +77,94 @@ type
...
+    handshakePattern: HandshakePattern
+    msgPatternIdx: uint8
+    psk: seq[byte]
+
+  # When a handshake is complete, the HandhshakeResult will contain the two 
+  # Cipher States used to encrypt/decrypt outbound/inbound messages
+  # The recipient static key rs and handshake hash values h are stored as well
+  HandshakeResult = object
+    cs1: CipherState
+    cs2: CipherState
```
[the remark this answers is not in the document]
Comment (s1fr0) [reply waku/v2/protocol/waku_noise/noise.nim:160]: Sure, this is a very reasonable change given also followup PR.

File: waku/v2/protocol/waku_noise/noise.nim [hunk 13]
```diff
@@ -81,9 +187,66 @@ type
...
+                              ),
+
+    "XX":     HandshakePattern(name: "Noise_XX_25519_ChaChaPoly_SHA256",
+                               preMessagePatterns: EmptyPreMessagePattern, 
+                               messagePatterns:    @[   MessagePattern(direction: D_r, tokens: @[T_e]),
+                                                        MessagePattern(direction: D_l, tokens: @[T_e, T_ee, T_s, T_es]),
+                                                        MessagePattern(direction: D_r, tokens: @[T_s, T_se])]
+                              ),
+
+    "XXpsk0": HandshakePattern(name: "Noise_XXpsk0_25519_ChaChaPoly_SHA256",
```
[the remark this answers is not in the document]
Comment (s1fr0) [reply waku/v2/protocol/waku_noise/noise.nim:222]: This is a custom handshake, used to immediately authenticate users based on some pre-shared key material (but not their static keys, since are exchanged again). It follows that security guarantees cannot be worse than XX, which has the same pattern except an extra pre-processing (with no data exchanged on the wire) of the `psk` information.

File: waku/v2/protocol/waku_noise/noise.nim [hunk 14]
```diff
@@ -62,6 +77,94 @@ type
...
+
+  # The pre message pattern consisting of a message direction and some Noise tokens, if any.
+  # (if non empty, only tokens e and s are allowed: http://www.noiseprotocol.org/noise.html#handshake-pattern-basics)
+  PreMessagePattern* = object
+    direction: MessageDirection
+    tokens: seq[NoiseTokens]
+
+  # The message pattern consisting of a message direction and some Noise tokens
+  # All Noise tokens are allowed
+  MessagePattern* = object
```
[the remark this answers is not in the document]
Comment (s1fr0) [reply waku/v2/protocol/waku_noise/noise.nim:109]: They are independent since `PreMessagePattern` [can only](http://www.noiseprotocol.org/noise.html#handshake-patterns) have patterns`e`, `s`, `e,s` or none. However, the token processing is completely similar among `PreMessagePattern` and `MessagePattern` , so I kept their internal structure (that is, direction + token) the same to reuse code (in followup PR).

In this way, "validity" of a `PreMessagePattern` can be checked with a external utility function, rather than complicating data structures.

File: waku/v2/protocol/waku_noise/noise.nim (hunk 9 above)
[the remark this answers is not in the document]
Comment (s1fr0) [reply waku/v2/protocol/waku_noise/noise.nim:541]: This check is now moved in encryptWithAd/decryptWithAd. See https://github.com/status-im/nwaku/pull/933#discussion_r875832985

File: waku/v2/protocol/waku_noise/noise.nim (hunk 12 above)
[the remark this answers is not in the document]
Comment (s1fr0) [reply waku/v2/protocol/waku_noise/noise.nim]: Addressed in https://github.com/status-im/nwaku/pull/933/commits/4ae4364845643190a3af13b3f15fd4afb6bcf906

File: waku/v2/protocol/waku_noise/noise.nim [hunk 15]
```diff
@@ -62,6 +77,94 @@ type
...
+    psk: seq[byte]
+
+  # When a handshake is complete, the HandhshakeResult will contain the two 
+  # Cipher States used to encrypt/decrypt outbound/inbound messages
+  # The recipient static key rs and handshake hash values h are stored as well
+  HandshakeResult = object
+    cs1: CipherState
+    cs2: CipherState
+    rs: EllipticCurveKey
+    h: MDigest[256] #The handshake state for channel binding
```
[the remark this answers is not in the document]
Comment (s1fr0) [reply waku/v2/protocol/waku_noise/noise.nim]: Addressed in https://github.com/status-im/nwaku/pull/933/commits/4ae4364845643190a3af13b3f15fd4afb6bcf906

File: waku/v2/protocol/waku_noise/noise.nim [hunk 16]
```diff
@@ -98,10 +261,333 @@ proc genKeyPair*(rng: var BrHmacDrbgContext): KeyPair =
...
+
+  # The nonce is read from the input CipherState
+  # By Noise specification the nonce is 8 bytes long out of the 12 bytes supported by ChaChaPoly
+  nonce[4..<12] = toBytesLE(state.n)
+
+  # Since ChaChaPoly decryption primitive overwrites the input with the output,
+  # we copy the ciphertext (authorization tag excluded) in the output plaintext variable and we pass it to decryption  
+  var plaintext = ciphertext[0..(ciphertext.high - ChaChaPolyTag.len)]
+  
+  ChaChaPoly.decrypt(state.k, nonce, authorizationTag, plaintext, ad)
```
[the remark this answers is not in the document]
Comment (s1fr0) [reply waku/v2/protocol/waku_noise/noise.nim]: Addressed in https://github.com/status-im/nwaku/pull/933/commits/4ae4364845643190a3af13b3f15fd4afb6bcf906

File: waku/v2/protocol/waku_noise/noise.nim (hunk 11 above)
[the remark this answers is not in the document]
Comment (s1fr0) [reply waku/v2/protocol/waku_noise/noise.nim]: Addressed in https://github.com/status-im/nwaku/pull/933/commits/4ae4364845643190a3af13b3f15fd4afb6bcf906

File: waku/v2/protocol/waku_noise/noise.nim (hunk 10 above)
[the remark this answers is not in the document]
Comment (s1fr0) [reply waku/v2/protocol/waku_noise/noise.nim]: Addressed in https://github.com/status-im/nwaku/pull/933/commits/4ae4364845643190a3af13b3f15fd4afb6bcf906

File: tests/v2/test_waku_noise.nim (hunk 6 above)
Comment (s1fr0) [reply tests/v2/test_waku_noise.nim:371]: Addressed in https://github.com/status-im/nwaku/pull/933/commits/4ae4364845643190a3af13b3f15fd4afb6bcf906

File: waku/v2/protocol/waku_noise/noise.nim (hunk 14 above)
[the remark this answers is not in the document]
Comment (staheri14) [reply waku/v2/protocol/waku_noise/noise.nim:109]: Is there a validator implemented for the `PreMessagePattern`? if not, please add a TODO in the code to as not to forget to implement, also please add the constraint you explained here i.e., `PreMessagePattern can only have patterns, s, e,s or none` as a comment for the `PreMessagePattern` data type.

File: tests/v2/test_waku_noise.nim (hunk 8 above)
[the remark this answers is not in the document]
Comment (staheri14) [reply tests/v2/test_waku_noise.nim:341]: I missed that part, right, the implicit is fine so is the explicit

File: waku/v2/protocol/waku_noise/noise.nim (hunk 7 above)
Comment (staheri14) [reply waku/v2/protocol/waku_noise/noise.nim]: thanks, clarified!

**Review by staheri14 — APPROVED** [review]
LGTM! (added a dew comments inline)

**Comment by s1fr0** [comment]
> I think you forgot to add the comment about this being optional?

@staheri14 I guess you're referring to https://github.com/status-im/nwaku/pull/933#discussion_r875283900, right?

If so, a comment was added here https://github.com/status-im/nwaku/blob/f5d4526cba5ab2ba904b62add830262955d78549/waku/v2/protocol/waku_noise/noise.nim#L157-L158 and here https://github.com/status-im/nwaku/blob/f5d4526cba5ab2ba904b62add830262955d78549/waku/v2/protocol/waku_noise/noise.nim#L162

File: waku/v2/protocol/waku_noise/noise.nim (hunk 15 above)
[the remark this answers is not in the document]
Comment (s1fr0) [reply waku/v2/protocol/waku_noise/noise.nim]: See https://github.com/status-im/nwaku/pull/933#issuecomment-1132537266

File: waku/v2/protocol/waku_noise/noise.nim (hunk 14 above)
[the remark this answers is not in the document]
Comment (s1fr0) [reply waku/v2/protocol/waku_noise/noise.nim:109]: By being related to handshake processing rules, such validator is already planned to be added in follow PR https://github.com/status-im/nwaku/pull/934.

File: tests/v2/test_waku_noise.nim (hunk 8 above)
[the remark this answers is not in the document]
Comment (s1fr0) [reply tests/v2/test_waku_noise.nim:331]: I personally prefer the implicit nonce setting, so I'll revert this change. Done in https://github.com/status-im/nwaku/pull/933/commits/1007980789a8e06813f8956663af03576f33d39b.

**Comment by s1fr0** [comment]
> ```
> expect NoiseNonceMaxError:
>       ciphertext = encryptWithAd(cipherState, ad, plaintext)
> ```

@staheri14 great suggestion, thanks! I changed unit tests accordingly in https://github.com/status-im/nwaku/pull/933/commits/bd11b37721166d16a399cfacc9ae5c357433f83d.

File: waku/v2/protocol/waku_noise/noise.nim (hunk 14 above)
[the remark this answers is not in the document]
Comment (s1fr0) [reply waku/v2/protocol/waku_noise/noise.nim:109]: Added in https://github.com/status-im/nwaku/pull/934/commits/59857903fc4ffed4a345fc40459fec1d38b3798a
